### ISSUE TITLE
Build a BertTextFieldEmbedder

### ISSUE THREAD

**Issue by valsworthen** [allenai/allennlp#2236]
Hello,

I am trying to use BERT for QA. In the paper, the authors explain that in this case they format the input data to be `Question tokens [SEP] Passage tokens`. Then a "mask" vector is also passed to distinguish between the question and the passage.

If I want to do the same using AllenNLP. I figured out from the code, that, when the `BertTokenEmbedder` is called [it can use the `token_type_ids`](https://github.com/allenai/allennlp/blob/1938a5af6d1ea3a2d8dccbb831f82990800fa3e1/allennlp/modules/token_embedders/bert_token_embedder.py#L89) which is this "mask". However, if in the config file we only specify 
```
"text_field_embedder": {
        "token_embedders": {
            "tokens": {
                "type": "bert-pretrained",
                "pretrained_model": "bert-base-uncased",
                "requires_grad": false,
                "top_layer_only": false
            }
        },
        "allow_unmatched_keys": true
}
```
[so the only key used by the TextFieldEmbedder is `'tokens'`](https://github.com/allenai/allennlp/blob/1938a5af6d1ea3a2d8dccbb831f82990800fa3e1/allennlp/modules/text_field_embedders/basic_text_field_embedder.py#L101). Then if I understand correctly, we should also pass a `embedder_to_indexer_map` to the TextFieldEmbedder.

To my point of view there remains the problem that [the `WordpieceIndexer` doesn't return the `token_type_ids`](https://github.com/allenai/allennlp/blob/1938a5af6d1ea3a2d8dccbb831f82990800fa3e1/allennlp/data/token_indexers/wordpiece_indexer.py#L165) so we should build it ourself.

In all cases I think this would imply that the DatasetReader returns Instances in the form `Question tokens [SEP] Passage tokens`.

Therefore I think that I need to:

- modify my Reader so that it returns Instances in the concatenated format (this way they would be indexed correctly). In this case how can correctly add the Token for [SEP] which is otherwise be tokenized as `[ SEP ]`?
- in its method `text_to_instance`, build the key `token_type_ids` and add it to the instance. edit: actually this should be done in a new TokenIndexer or in the Model
- add the following `'tokens': 'token_type_ids'` to the config file under the key `"text_field_embedder"` and pass a map.

Can you tell me if my understanding of the different parts is correct and if my corrections would work? 

Many thanks for your help!

**Comment by WrRan** [comment]
I am working on it. Hope my solution (https://github.com/WrRan/allennlp/tree/enhance-bert) helps you.

I modify two allennlp library files:
`allennlp/data/token_indexers/wordpiece_indexer.py`(https://github.com/WrRan/allennlp/blob/6d42b903f2d60148e4f8c7ca387889ed0e6fa49b/allennlp/data/token_indexers/wordpiece_indexer.py#L255)
```py
class WordpieceIndexer(TokenIndexer[int]):
    @overrides
    def tokens_to_indices(self,
                          tokens: List[Token],
                          vocabulary: Vocabulary,
                          index_name: str) -> Dict[str, List[int]]:
        # some codes go here...
        # By construction, we still have enough room to add the end_token ids.
        wordpiece_ids.extend(self._end_piece_ids)
        # Constructing `token_type_ids` by `self._end_piece_ids`
        token_type_ids = _get_token_type_ids(wordpiece_ids,
                                             self._end_piece_ids)

        # Our mask should correspond to the original tokens,
        # because calling util.get_text_field_mask on the
        # "wordpiece_id" tokens will produce the wrong shape.
        mask = [1 for _ in tokens]

        return {
                index_name: wordpiece_ids,
                f"{index_name}-offsets": offsets,
                f"{index_name}-type-ids": token_type_ids,
                "mask": mask
        }

def _get_token_type_ids(wordpiece_ids: List[int],
                        end_piece_ids: List[int]) -> List[int]:
    length_wordpiece = len(wordpiece_ids)
    length_end_pieces = len(end_piece_ids)
    token_type_ids = [0] * length_wordpiece
    type_id = 0
    cursor = 0
    while cursor < len(wordpiece_ids):
        if end_piece_ids[0] != wordpiece_ids[cursor]:
            cursor += 1
        else:
            if end_piece_ids == wordpiece_ids[cursor:cursor+length_end_pieces]:
                type_id += 1
                cursor += length_end_pieces
                token_type_ids[cursor:] = [type_id] * (length_wordpiece - cursor)
            else:
                cursor += length_end_pieces
    return token_type_ids
```
and `allennlp/modules/text_field_embedders/basic_text_field_embedder.py`(https://github.com/WrRan/allennlp/blob/6d42b903f2d60148e4f8c7ca387889ed0e6fa49b/allennlp/modules/text_field_embedders/basic_text_field_embedder.py#L96):
```py
            # If we pre-specified a mapping explictly, use that.
            # If `indexer_key` is None, we map it to `None`.
            if self._embedder_to_indexer_map is not None:
                tensors = [(text_field_input[indexer_key] if indexer_key is not None else None)
                           for indexer_key in self._embedder_to_indexer_map[key]]
```

Then, you can config it like this to use it (do not need `tokens-offsets`):
```json
        "text_field_embedder": {
          "token_embedders": {
            "tokens": {
              "type": "bert-pretrained",
              "pretrained_model": "bert-base-uncased",
              "requires_grad": true
            }
          },
          "embedder_to_indexer_map": {
            "tokens": ["tokens", null, "tokens-type-ids"]
          },
          "allow_unmatched_keys": true
        }
```
or (need `tokens-offsets`):
```json
        "text_field_embedder": {
          "token_embedders": {
            "tokens": {
              "type": "bert-pretrained",
              "pretrained_model": "bert-base-uncased",
              "requires_grad": true
            }
          },
          "embedder_to_indexer_map": {
            "tokens": ["tokens", "tokens-offsets", "tokens-type-ids"]
          },
          "allow_unmatched_keys": true
        }
```

**Comment by SparkJiao** [comment]
@WrRan Very appreciate to your work, will this be merged into soon?

**Comment by WrRan** [comment]
I am willing to create a PR.
But before that, @joelgrus do you have a suggestion or a better solution?

**Comment by joelgrus** [comment]
here's what I would probably do:

1. change the wordpiece indexer to (always) return token_type_ids
2. in the wordpiece indexer constructor, take an optional `separator: str = "[SEP]"` character
3. use that separator to construct the appropriate token_type_ids
4. have your dataset reader create a text field that has both sentences and the separator like you did

but in this case I'm not sure what to do about sequences that are too long, the current logic won't work.

(also, your change to textfieldembedder is unnecessary, instead of doing

```
"tokens": ["tokens", null, "tokens-type-ids"]
```

you should be able to just do

```
"tokens": ["tokens", "tokens-type-ids"]
```

it's the names that matter, not the positions)

does that make sense?

**Comment by SparkJiao** [comment]
For longer sequence, may be you could see this: [27](https://github.com/google-research/bert/issues/27) ?
We could make a scalar mix for the boundary words or choose the hidden state where it has the longest context ? But I think both will be a difficult work while implementing...

**Comment by joelgrus** [comment]
I see, the `embedder(*tensors)` is really unfortunate, I wish it were doing something like

```
         tensors = {indexer_key: text_field_input[indexer_key] for 
                    indexer_key in self._embedder_to_indexer_map[key]}
```

and then

```
embedder(**tensors)
```

but I'm sure that would break things. let me think about it, I don't like having to stick nulls in the `embedder_to_indexer_map`, it seems fragile and confusing, but I'm not sure if there's a better way.

**Comment by valsworthen** [comment]
@WrRan thanks for your work, I think this solves the problem! However, regarding the dataset reader, don't you think that it also needs to add a `[CLS]` token at the beginning and another `[SEP]` token at the end of the example? I think this is done by Google and HuggingFace: [see here](https://github.com/huggingface/pytorch-pretrained-BERT/blob/8da280ebbeca5ebd7561fd05af78c65df9161f92/examples/extract_features.py#L82) and [there](https://github.com/huggingface/pytorch-pretrained-BERT/blob/8da280ebbeca5ebd7561fd05af78c65df9161f92/examples/run_squad.py#L237).

Also I need to check if the spans are calculated using Wordpiece index of the "real" token index.

**Comment by WrRan** [comment]
> I see, the `embedder(*tensors)` is really unfortunate, I wish it were doing something like
> 
> ```
>          tensors = {indexer_key: text_field_input[indexer_key] for 
>                     indexer_key in self._embedder_to_indexer_map[key]}
> ```
> and then
> 
> ```
> embedder(**tensors)
> ```
> but I'm sure that would break things. let me think about it, I don't like having to stick nulls in the `embedder_to_indexer_map`, it seems fragile and confusing, but I'm not sure if there's a better way.

I have supported what you said in PR #2279. Do you have any suggestions?

**Comment by matt-gardner** [comment]
Fixed by #2279.
